packages/protocol/test/governance/validators.ts

```diff
@@ -2178,10 +2178,145 @@ contract('Validators', (accounts: string[]) => {
       })
     })
 
-    describe('when the sender is not an approved address', async () => {
+    describe('when the sender is not an approved address', () => {
       it('should revert', async () => {
         await assertRevert(validators.forceDeaffiliateIfValidator(validator))
       })
     })
   })
+
+  describe('#groupMembershipInEpoch', () => {
+    const validator = accounts[0]
+    const groups = accounts.slice(1, -1)
+
+    beforeEach(async () => {
+      await registry.setAddressFor(CeloContractName.DowntimeSlasher, validator)
+      await registry.setAddressFor(CeloContractName.DoubleSigningSlasher, accounts[3])
+      await registry.setAddressFor(CeloContractName.Governance, accounts[4])
+    })
+
+    describe('when the validator is added to different groups sequentially', () => {
+      const epochs = []
+      beforeEach(async () => {
+        await registerValidator(validator)
+        for (const group of groups) {
+          await registerValidatorGroup(group)
+        }
+        for (let i = 0; i < membershipHistoryLength.plus(1).toNumber(); i++) {
+          const blockNumber = await web3.eth.getBlockNumber()
+          const epochNumber = Math.floor(blockNumber / EPOCH)
+          const blocksUntilNextEpoch = (epochNumber + 1) * EPOCH - blockNumber
+          await mineBlocks(blocksUntilNextEpoch, web3)
+          epochs.push([epochNumber + 1, i !== 4 ? groups[i] : NULL_ADDRESS])
+
+          if (i !== 4) {
+            await validators.affiliate(groups[i])
+            await validators.addFirstMember(validator, NULL_ADDRESS, NULL_ADDRESS, {
+              from: groups[i],
+            })
+          } else {
+            await validators.forceDeaffiliateIfValidator(validator)
+          }
+        }
+      })
+
+      it('should always return the correct membership for the given epoch', async () => {
+        for (let i = 0; i < membershipHistoryLength.toNumber(); i++) {
+          assert.equal(
+            await validators.groupMembershipInEpoch(validator, epochs[i + 1][0], 2 + i),
+            i !== 3 ? epochs[i + 1][1] : NULL_ADDRESS
+          )
+        }
+      })
+    })
+
+    describe('when the validator is added to different groups with gaps in between epochs', () => {
+      let epochs = []
+
+      beforeEach(async () => {
+        epochs = []
+        await registerValidator(validator)
+        for (const group of groups) {
+          await registerValidatorGroup(group)
+        }
+        for (let i = 0; i < membershipHistoryLength.multipliedBy(3).toNumber(); i++) {
+          const blockNumber = await web3.eth.getBlockNumber()
+          const epochNumber = Math.floor(blockNumber / EPOCH)
+          const blocksUntilNextEpoch = (epochNumber + 1) * EPOCH - blockNumber
+          await mineBlocks(blocksUntilNextEpoch, web3)
+
+          if (i % 3 === 0) {
+            epochs.push([epochNumber + 1, i !== 9 ? groups[i / 3] : NULL_ADDRESS])
+            if (i !== 9) {
```

Comment:
Why do this?

Reply:
The reason behind `i % 3 == 0` was to only add `membershipHistoryLength` entries - the reason was to avoid the confusion of only tracking the last `membershipHistoryLength` entries while still adding epochNumber gaps between entries. 
The reason behind `i !== 9` is to add a deaffiliation _between_ other entries. 
I'll definitely change this to be a function of `membershipHistoryLength` rather than magic numbers 3 and 4.
If you want me to do this differently, let me know and I can look into it.